src/contracts/core/AllocationManager.sol

```diff
@@ -198,13 +199,77 @@ contract AllocationManager is
+            // Check the operator set exists and the operator is not currently slashable
+            OperatorSet memory operatorSet = OperatorSet(params.avs, params.operatorSetIds[i]);
+            require(_operatorSets[operatorSet.avs].contains(operatorSet.id), InvalidOperatorSet());
+            require(!isOperatorSlashable(msg.sender, operatorSet), AlreadyMemberOfSet());
+
+            // Add operator to operator set
+            registeredSets[msg.sender].add(operatorSet.key());
+            _operatorSetMembers[operatorSet.key()].add(msg.sender);
+            emit OperatorAddedToOperatorSet(msg.sender, operatorSet);
+
+            registrationStatus[msg.sender][operatorSet.key()].registered = true;
+        }
+
+        // Call the AVS to complete registration. If the AVS reverts, registration will fail
+        IAVS(params.avs).registerOperator(msg.sender, params.operatorSetIds, params.data);
+    }
+
+    /// @inheritdoc IAllocationManager
+    function deregisterFromOperatorSets(
+        DeregisterParams calldata params
+    ) external onlyWhenNotPaused(PAUSED_OPERATOR_SET_REGISTRATION_AND_DEREGISTRATION) {
+        require(msg.sender == params.operator || msg.sender == params.avs, InvalidCaller());
+
+        // Check that the operator exists
+        require(delegation.isOperator(params.operator), InvalidOperator());
+
+        for (uint256 i = 0; i < params.operatorSetIds.length; i++) {
+            // Check the operator set exists and the operator is a member
+            OperatorSet memory operatorSet = OperatorSet(params.avs, params.operatorSetIds[i]);
+            require(_operatorSets[params.avs].contains(operatorSet.id), InvalidOperatorSet());
+            require(registrationStatus[params.operator][operatorSet.key()].registered, NotMemberOfSet());
+            require(isOperatorSlashable(params.operator, operatorSet), NotMemberOfSet());
+
+            // Remove operator from operator set
+            registeredSets[params.operator].remove(operatorSet.key());
+            _operatorSetMembers[operatorSet.key()].remove(params.operator);
+            emit OperatorRemovedFromOperatorSet(params.operator, operatorSet);
+
+            // The operator is deregistered, but remains slashable until the deallocation delay
+            // passes. Once this delay passes, the operator can instantly deallocate any stake
+            // allocated to this operator set.
+            // forgefmt: disable-next-item
+            registrationStatus[params.operator][operatorSet.key()] = RegistrationStatus({
+                registered: false,
+                slashableUntil: uint32(block.number) + DEALLOCATION_DELAY
+            });
+
+            // TODO - should this instead do modifyAllocations + removing allocations?
+            // issue right now is if I call deregister, i have to actually go remove allocations.
+            // if i re-register, i immediately become slashable again
+        }
```

Comment:
Heavily in favor of removing allocations once you deregister. 

If we go down the route of not removing allocations, we still need to enable proper introspection with something like: 

```solidity
struct Registration Status{
    bool registered;
    bool allocated;
    uint32 slashableUntil;
}
```

Reply:
ok that's what i was feeling too. lemme take a crack at it today...

Reply:
I guess the main question here is how we handle pending modifications.

Obviously the main case we want to handle is when we have existing magnitude, and no pending modification. Easy - create a pending deallocation for the current magnitude.

What about these cases, though?

1. `(curMagnitude: 0, pendingDiff > 0)`
2. `(curMagnitude > 0, pendingDiff > 0)`
3. `(curMagnitude > 0, pendingDiff < 0)`

Reply:
I am opposed to removing allocations upon deregistration, primarily due to the case of ejection. Removing allocations upon ejection means it takes at least `DEALLOCATION_DELAY+allocationDelay(op)` in order to rejoin the operatorSet.

EigenDA recently made its ejection cooldown 3 days instead of 7 days so that ejected operators can join sooner.

Reply:
[cacf8f9](https://github.com/Layr-Labs/eigenlayer-contracts/pull/860/commits/cacf8f97c1bb307bc45b1d8e103a7dadd8bcef37)

removed deallocations on deregistration here